pkg/addon-operator/ensure_crds.go

```diff
@@ -0,0 +1,225 @@
+type CRDsInstaller struct {
+	k8sClient    dynamic.Interface
+	crdFilesPath []string
+	buffer       []byte
+
+	// concurrent tasks to create resource in a k8s cluster
+	k8sTasks *multierror.Group
+}
+
+func (cp *CRDsInstaller) Run(ctx context.Context) *multierror.Error {
+	result := new(multierror.Error)
+
+	for _, crdFilePath := range cp.crdFilesPath {
+		if match := strings.HasPrefix(filepath.Base(crdFilePath), "doc-"); match {
+			continue
+		}
+
+		err := cp.processCRD(ctx, crdFilePath)
+		if err != nil {
+			err = fmt.Errorf("error occurred during processing %q file: %w", crdFilePath, err)
+			result = multierror.Append(result, err)
+			continue
+		}
+	}
+
+	errs := cp.k8sTasks.Wait()
+	if errs.ErrorOrNil() != nil {
+		result = multierror.Append(result, errs.Errors...)
+	}
+
+	return result
+}
+
+func (cp *CRDsInstaller) processCRD(ctx context.Context, crdFilePath string) error {
+	crdFileReader, err := os.Open(crdFilePath)
+	if err != nil {
+		return err
+	}
+	defer crdFileReader.Close()
+
+	crdReader := apimachineryYaml.NewDocumentDecoder(crdFileReader)
+
+	for {
+		n, err := crdReader.Read(cp.buffer)
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+
+			return err
+		}
+
+		data := cp.buffer[:n]
+		if len(data) == 0 {
+			// some empty yaml document, or empty string before separator
+			continue
+		}
+		rd := bytes.NewReader(data)
+		err = cp.putCRDToCluster(ctx, rd, n)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (cp *CRDsInstaller) putCRDToCluster(ctx context.Context, crdReader io.Reader, bufferSize int) error {
+	var crd *v1.CustomResourceDefinition
+
+	err := apimachineryYaml.NewYAMLOrJSONDecoder(crdReader, bufferSize).Decode(&crd)
+	if err != nil {
+		return err
+	}
+
+	// it could be a comment or some other peace of yaml file, skip it
+	if crd == nil {
+		return nil
+	}
+
+	if crd.APIVersion != v1.SchemeGroupVersion.String() && crd.Kind != "CustomResourceDefinition" {
+		return fmt.Errorf("invalid CRD document apiversion/kind: '%s/%s'", crd.APIVersion, crd.Kind)
+	}
+
+	if len(crd.ObjectMeta.Labels) == 0 {
+		crd.ObjectMeta.Labels = make(map[string]string, 1)
+	}
+	crd.ObjectMeta.Labels["heritage"] = "deckhouse"
+
+	cp.k8sTasks.Go(func() error {
+		return cp.updateOrInsertCRD(ctx, crd)
+	})
+
+	return nil
+}
+
+func (cp *CRDsInstaller) updateOrInsertCRD(ctx context.Context, crd *v1.CustomResourceDefinition) error {
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		existCRD, err := cp.getCRDFromCluster(ctx, crd.GetName())
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				ucrd, err := sdk.ToUnstructured(crd)
+				if err != nil {
+					return err
+				}
+
+				_, err = cp.k8sClient.Resource(crdGVR).Create(ctx, ucrd, apimachineryv1.CreateOptions{})
+				return err
+			}
+
+			return err
+		}
+
+		if existCRD.Spec.Conversion != nil {
+			crd.Spec.Conversion = existCRD.Spec.Conversion
+		}
+
+		if existCRD.GetObjectMeta().GetLabels()["heritage"] == "deckhouse" &&
```

Comment:
Let's remove any mentioning of deckhouse product. As I understand there was an idea to add some extra env to the addon-operator like ADDON_OPERATOR_EXTRA_LABELS with default value `heritage=addon-operator`. This way it would be possible to redefine this env if needed to something like `heritage=not-addon-operator,foo=bar` and apply custom `heritage` value to modules` CRDs.

Reply:
DONE

Reply:
DONE